### ISSUE TITLE
Standardizing Github Teams

### ISSUE THREAD

**Issue by nonword** [NYPL/engineering-general#51]
*tldr; Teams are a mess. By project maybe?*

Our Github Teams situation is a mess. Our teams include a mix of teams created for individual projects and teams reflecting prior institutaional hierarchy. I wasn't able to find useful online opinion on patterns for using Github Teams to simplify repository access across multiple interrelated projects & repos. Our challenge is to 1) maximize the ease with which we can grant `Read` and `Write` access to relevant staff while simultaneously 2) withholding access that is unwanted, unnecessary, or dangerous. We want things locked-down in general, but we're willing to be a little permissive with permissions if it means we save time tracking down owners/admins to manage account-specific access.

As an aside, I think our repositories should be public by default. If we're following our own recommendations, there's no danger making them so. This simplifies `Read` access concerns, although public repos would still benefit from better organized Teams for granting `Write` access.

The major patterns of Team organization seem to be:

## I. Recreate institutional hierarchy (i.e. group by role)

The vanilla use case favored by Github's [Nested teams documentation](https://github.com/blog/2378-nested-teams-add-depth-to-your-team-structure) doesn't seem well suited to our environment on its own. I don't think we'd benefit from teams solely organized by role like "NYPL Digital Engineers", "NYPL Digital PMs", "NYPL Digital Design/UX", etc. We don't often need to grant read/write access only to Developers - and to all Developers. In practice our projects involve *certain* developers, only one PM, and a selection of Design/UX, and subject specialists - some outside Digital.

## II. One team per project

Our work is typically organized by "project", a thematically squashed collection of products supported by a number of repositories. A project tends to have a regular stand-up and a dedicated Jira/Waffle board. Contributors to a project often benefit from having unified access to all of the repositories involved in that project. Thus, one pattern we could adopt is to create Github teams around projects. At the start of a project, for the purpose of easily giving people read/write access to the repositories for a project, a Github team could be created that matches the product name (e.g. "Best Books", "Discovery/ReCAP", "SimplyE"). The team would include every Github account that might be interested in seeing the contents of the repo. A more selective "Write-Access" sub-team could be created to represent those contributors likely to need write access to the majority of the repositories associated with the project (e.g. "Best Books Write-Access", "Discovery/ReCAP Write-Access"). Anyone added to the project would need only be added to the relevant project team to gain read/write access to all relevant repositories.

(I really don't like "Write-Access" as a suffix, but "Developer"/"Contributor" are presumptuous/limiting.)

Github project teams would not be deleted when project work is complete because doing so may revoke read/write access to the effected repositories, which may make troubleshooting issues post-launch difficult/impossible. Thus, under this model, the list of Github teams is expected to increase steadily forever and that's okay. Teams are free and their persistence may be valuable as a kind of rolling secondary documentation of repository-project relationships (in lieu of using Github's "Projects" feature).

Note that a given repository may be implicated in multiple projects and thus may grant `Read`/`Write` access to multiple Teams. For example, the repositories representing the Inventory Service may grant `Write` access to both "Discovery/ReCAP Write-Access" and "New Arrivals Write-Access".

For **common repositories** not tied to a specific project (e.g. NYPL/nypl-data-api-client, NYPL/nypl-core, NYPL/engineering-general, etc.), we would be forced to make them Public or grant `Read` access to a "NYPL" catch-all team and `Write` access to either:
 a) individual accounts on a case by case basis, or
 b) role-based teams like "NYPL Engineers", "NYPL Metadata Services".

## III. Grant `Read` to "NYPL", `Write` by project

This option abandons granular, project-specific `Read` Teams in favor of simply granting `Read` to every member of a catch-all "NYPL" team. We gain the convenience of not having to manage project or role-specific teams for `Read` access. The cost of this is that every member of the "NYPL" catch-all team gets `Read` access to every NYPL private repo. (I see no danger in that provided we're not storing secrets.) To retain granular control over `Write` access, project specific Teams would be created (e.g. "Best Books Write-Access", "Discovery/ReCAP Write-Access").

Our Teams listing might resemble:

 * NYPL
    * NYPL Engineers
    * NYPL Metadata Services
    * NYPL Design/UX
 * Best Books Write-Access
 * SimplyE Write-Access

For **common repositories** not tied to a specific project, our options resemble those in II.

## Questions

1. What option or mix of options above maximizes convenience and security for us?
2. What do we do about `Admin` access, which by default is held solely by repository creators. If we create project based teams, the only time we should need `Admin` access is when initially granting team access to the repo, so perhaps that power is sufficient shared solely by the repo owner and NYPL org admin.

**Comment by nkrabben** [comment]
Popping in to mention that non-Digital staff also have repos on Github. The [Special Collections team](https://github.com/orgs/NYPL/teams/special-collections) will have to discuss what policy we want to use to manage read and write access. However, our repos don't currently overlap with any active projects in Digital.

**Comment by nodanaonlyzuul** [comment]
### Regarding Public Repos by default

I'm okay with that. Sometimes I think that, 
a casual, external Github user would get confused or overwhelmed by the number of repos we have.
But I'm not sure if other people have the expectation that these repos be for their consumption.

As you point out it would really simplify the `Read` and `Write` persmissions.
If someone who isn't a regular contributor wants to commit, they'll be able to fork & file a PR from their fork.
As of today - they're locked out period.


### Option 3 is appealing

I'm also fine with the fact that everyone in the "catch all" team  will be set 
as watchers whenever a repo is created. Right? They'll just have to either un-watch
or get a lot of email.

We have a long history of errantly or intentionally committing of secrets.
I'm mostly fine letting those repos stay private and still adding the "general" team to those.
Not in a hippy-dippy sense that I trust former employees. More that a former employee's access
to secrets is hard to lock down and their misuse of those secrets is a criminal act.
It's a person problem - as much as a computer problem.

Thank you for getting this conversation going (again!)

**Comment by nonword** [comment]
@nkrabben True and thanks for highlighting. Any standard that comes out of this needn't extend beyond those agreeing to it. It may just be me.

@nodanaonlyzuul Yeah the Public/Private is kind of a red herring and I definitely don't want to make any bulk changes of that sort to existing repos. But thought it worth noting given that open-by-default simplifies permissions, which is the driving interest of this Teams work. As for automatically watching repos, I'd be buried without having unchecked `Settings > Notifications > Automatically watch repositories` long ago, but yeah maybe a lot of people would be spammed by this change.

**Comment by seanredmond** [comment]
I just unchecked `Automatically watch repositories` so I already learned something important from this discussion. Thank you @nonword ! 

One of the problems we keep having is that people just can't find the repos they need because they can't see them at all. I like granting broad `Read` access and spreading the knowledge of how people can keep down the volume of unwanted messages.

**Comment by asyroyez** [comment]
Would outside help be welcome? I used to work for NYPL but quit semi-recently to study code. I am not sure how much I can contribute over the next few months while doing a bootcamp, but I wanted to express a willingness!

> Sometimes I think that,
a casual, external Github user would get confused or overwhelmed by the number of repos we have.
@nodanaonlyzuul 

Very correct... especially for a newbie! I kind of get the gist from working there previously, but would appreciate some help with where to look first.

**Comment by nodanaonlyzuul** [comment]
1. @asyroyez - congrats on the career change!
2. I'm with @seanredmond - @nonword, I didn't know that feature existed - thank you for adding minutes to my day and storage to my inbox. You've assuaged any issue I'd have with making those public. Tonight, I celebrate my love for you.

Finally - organizations have a "description" section.
We can use that to link to an entry in [NYPL/engineering-general](https://github.com/NYPL/engineering-general) where we discuss why we develop in the open.
We can use that as a space to:

1.  Hedge expectation about our internal apps' usefulness outside NYPL.
2.  Do some wayfinding for outsiders, for projects that useful-to/ripe-for non-employees.

**Comment by mattkoch614** [comment]
This issue is quite old, over a year. Closing, with a tag of `archived`.
